File: Python/perfect_squares.py
class Solution():
    def __init__(self):
        self.name = 'S->13'

    # Time complexity: O(nlogn)
    # Space complexity: O(n)
    def SumOfSquares(self, n):
        i = 1
        squares = []
        while i * i <= n:
            squares.append(i*i)
            i += 1

        min_sums = [n] *  (n+1)
        min_sums[0] = 0
        for i in range(0, n+1):
            for s in squares:
                if i + s < n+1:
                    min_sums[i+s] = min(min_sums[i+s], min_sums[i] + 1)
        return min_sums[-1]
n = 4
print(Solution().SumOfSquares(n))

File: Javascript/perfect_squares.js
// Time complexity: O(nlogn)
// Space complexity: O(n)
var numSquares = function(n) {
    var i = 0;
    squares = []
    while (i * i <= n) {
        squares.push(i * i)
        i++
    }
    
    min_sums = new Array(n+1).fill(n)
    min_sums[0] = 0
    for (var i = 0; i <= n; ++i) {
        for (let s of squares) {
            if (i + s <= n) {
                min_sums[i + s] = Math.min(min_sums[i + s], min_sums[i] + 1)
            }
        }
    }
    return min_sums[min_sums.length-1]
};
